src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1295 @@
+      }
+    }
+    // In case we had previously assigned this token to be reprocessed, we can
+    // erase it from that set because it's "happy now" (we just processed it).
+    reprocess.erase(tok);
+  }
+
+  size_t max_loop = 1000000, loop_count; // max_loop is to detect epsilon cycles.
+  for (loop_count = 0; !reprocess.empty() && loop_count < max_loop; ++loop_count) {
+    std::vector<Token *> reprocess_vec;
+    for (typename unordered_set<Token *>::iterator iter = reprocess.begin();
+         iter != reprocess.end(); ++iter)
+      reprocess_vec.push_back(*iter);
+    reprocess.clear();
+    for (typename std::vector<Token *>::iterator iter = reprocess_vec.begin();
+         iter != reprocess_vec.end(); ++iter) {
+      Token *tok = *iter;
+      int32 pos = token2pos[tok];
+      // Repeat the processing we did above (for comments, see above).
+      for (ForwardLinkT *link = tok->links; link != NULL; link = link->next) {
+        if (link->ilabel == 0) {
+          IterType following_iter = token2pos.find(link->next_tok);
+          if (following_iter != token2pos.end()) {
+            int32 next_pos = following_iter->second;
+            if (next_pos < pos) {
+              following_iter->second = cur_pos++;
+              reprocess.insert(link->next_tok);
+            }
+          }
+        }
+      }
+    }
+  }
+  KALDI_ASSERT(loop_count < max_loop &&
+               "Epsilon loops exist in your decoding "
+               "graph (this is not allowed!)");
+
+  topsorted_list->clear();
+  topsorted_list->resize(cur_pos, NULL); // create a list with NULLs in between.
+  for (IterType iter = token2pos.begin(); iter != token2pos.end(); ++iter)
+    (*topsorted_list)[iter->second] = iter->first;
+}
+
+template <typename FST, typename Token>
+bool LatticeIncrementalDecoderTpl<FST, Token>::GetLattice(bool use_final_probs,
+                                                          bool redeterminize,
+                                                          int32 last_frame_of_chunk,
+                                                          CompactLattice *olat) {
+  using namespace fst;
+  bool not_first_chunk = last_get_lattice_frame_ != 0;
+
+  // last_get_lattice_frame_ is used to record the first frame of the chunk
+  // last time we obtain from calling this function. If it reaches NumFramesDecoded()
+  // we cannot generate any more chunk
+  if (last_get_lattice_frame_ < last_frame_of_chunk) {
+    Lattice raw_fst;
+    // step 1: Get lattice chunk with initial state
+    // In this function, we do not create the initial state in
+    // the first chunk, and we do not create the final state in the last chunk
+    KALDI_ASSERT(GetRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_,
+                               last_frame_of_chunk, not_first_chunk,
+                               !decoding_finalized_));
+    // step 2: Determinize the chunk
+    CompactLattice clat;
+#if 1
+    if (!DeterminizeLatticePhonePrunedWrapper(
```

Comment:
@danpovey Another problem I found is that only if we use a beam larger than beam+lattice_beam here, we can guarantee no final or initial arcs are pruned by this function (I already use extra_cost and tot_cost as final and initial arc weights respectively). 
These pruned final arcs can significantly hurt oracle WER performance but not 1-best WER (also result in less lattice density).
Moreover, if I use beam here, the oracle WER performance can be similar to the baseline (also similar lattice density).
Hence, do you prefer using (lattice_beam+beam) or (lattice_beam) or (beam) here?

In summary, after fixing these 3 problems (this problem,  "same state label multiple final arcs" and "distinguish final and initial arcs" I discuss before this conversation), we can get the same 1-best hypotheses in a testset subset.